File: api/hn_app/services/hn_service.py
import asyncio
import json
import threading
from datetime import datetime, timezone

import httpx
from httpx import HTTPError

BASE_URL = "https://hacker-news.firebaseio.com/v0"
STORY_CACHE: dict[int, dict] = {}
CACHE_TTL_SECONDS = 120
_cache_timestamp: float = 0.0
_cache_lock = threading.Lock()
CONCURRENCY_LIMIT = 10


def _clear_expired_cache() -> None:
    global _cache_timestamp
    with _cache_lock:
        now = datetime.now(tz=timezone.utc).timestamp()
        if now - _cache_timestamp > CACHE_TTL_SECONDS:
            STORY_CACHE.clear()
            _cache_timestamp = now


async def fetch_top_story_ids(client: httpx.AsyncClient) -> list[int]:
    resp = await client.get(f"{BASE_URL}/topstories.json")
    resp.raise_for_status()
    try:
        return resp.json()
    except (json.JSONDecodeError, ValueError):
        return []


async def fetch_story(client: httpx.AsyncClient, story_id: int, sem: asyncio.Semaphore) -> dict | None:
    async with sem:
        if story_id in STORY_CACHE:
            return STORY_CACHE[story_id]

        try:
            resp = await client.get(f"{BASE_URL}/item/{story_id}.json")
            resp.raise_for_status()
            try:
                data: dict | None = resp.json()
            except (json.JSONDecodeError, ValueError):
                return None
            if data and data.get("type") == "story" and data.get("title"):
                STORY_CACHE[story_id] = data
                return data
        except HTTPError:
            pass
        return None


async def get_top_stories(limit: int = 30) -> list[dict]:
    _clear_expired_cache()

    sem = asyncio.Semaphore(CONCURRENCY_LIMIT)

    async with httpx.AsyncClient() as client:
        ids = await fetch_top_story_ids(client)

        tasks = [fetch_story(client, sid, sem) for sid in ids[:limit]]
        results = await asyncio.gather(*tasks)

    items = [r for r in results if r is not None]

    items.sort(key=lambda x: x.get("time", 0), reverse=True)
    return items

File: src/services/hn-service.ts
import type { NewsItem } from "../schemas/news";

const BASE_URL = "https://hacker-news.firebaseio.com/v0";
const CACHE_TTL_MS = 120_000;

interface HnItem {
  id: number;
  title?: string;
  score?: number;
  by?: string;
  time?: number;
  url?: string;
  text?: string;
  descendants?: number;
  type?: string;
}

const cache = new Map<number, HnItem>();
let cacheTimestamp = 0;

function isCacheExpired(): boolean {
  return Date.now() - cacheTimestamp > CACHE_TTL_MS;
}

function clearExpiredCache(): void {
  if (isCacheExpired()) {
    cache.clear();
    cacheTimestamp = Date.now();
  }
}

async function fetchJson<T>(url: string): Promise<T> {
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`HTTP ${resp.status} fetching ${url}`);
  return resp.json() as Promise<T>;
}

async function fetchStory(id: number): Promise<HnItem | null> {
  const cached = cache.get(id);
  if (cached) return cached;

  try {
    const item = await fetchJson<HnItem>(`${BASE_URL}/item/${id}.json`);
    if (item && item.type === "story" && item.title) {
      cache.set(id, item);
      return item;
    }
  } catch {
    /* skip failed items */
  }
  return null;
}

export async function getTopStories(limit = 30): Promise<NewsItem[]> {
  clearExpiredCache();

  const ids = await fetchJson<number[]>(`${BASE_URL}/topstories.json`);
  const items = await Promise.all(ids.slice(0, limit).map(fetchStory));

  const valid = items.filter(Boolean) as HnItem[];
  valid.sort((a, b) => (b.score ?? 0) - (a.score ?? 0));

  return valid.map((item) => ({
    id: item.id,
    title: item.title ?? "Untitled",
    points: item.score ?? 0,
    author: item.by ?? "unknown",
    published: new Date((item.time ?? 0) * 1000).toISOString().replace("T", " ").replace(/\.\d{3}Z/, " UTC"),
    url: item.url ?? null,
    text: item.text ?? null,
    comments: item.descendants ?? 0,
  }));
}
